quark/mdk_runtime.q

```diff
@@ -0,0 +1,174 @@
+	Actor getScheduleService() {
+	    return ?self.dependencies.getService("schedule");
+	}
+
+    }
+
+    @doc("""
+    Return current time.
+    """)
+    interface Time {
+	@doc("""
+        Return the current time in seconds since the Unix epoch.
+        """)
+	float time();
+    }
+
+    @doc("""
+    An actor that can schedule events.
+
+    Accepts Schedule messages and send Happening events to originator at the
+    appropriate time.
+    """)
+    interface SchedulingActor extends Actor {}
+
+    @doc("""
+    Please send me a Happening message with given event name in given number of
+    milliseconds.
+    """)
+    class Schedule {
+	String event;
+	float seconds;
+
+	Schedule(String event, float seconds) {
+	    self.event = event;
+	    self.seconds = seconds;
+	}
+    }
+
+    @doc("A scheduled event is now happening.")
+    class Happening {
+	String event;
+	float currentTime;
+
+	Happening(String event, float currentTime) {
+	    self.event = event;
+	    self.currentTime = currentTime;
+	}
+    }
+
+    class _ScheduleTask extends Task {
+	QuarkRuntimeTime timeService;
+	Actor requester;
+	String event;
+
+	_ScheduleTask(QuarkRuntimeTime timeService, Actor requester, String event) {
+	    self.timeService = timeService;
+	    self.requester = requester;
+	    self.event = event;
+	}
+
+	void onExecute(Runtime runtime) {
+	    timeService.dispatcher.tell(
+	        self.timeService, new Happening(self.event, self.timeService.time()), self.requester);
+	}
+    }
+
+    @doc("""
+    Temporary implementation based on Quark runtime, until we have native
+    implementation.
+    """)
+    class QuarkRuntimeTime extends Time, SchedulingActor {
+	MessageDispatcher dispatcher;
+
+	void onStart(MessageDispatcher dispatcher) {
+	    self.dispatcher = dispatcher;
+	}
+
+	void onMessage(Actor origin, Object msg) {
+	    Schedule sched = ?msg;
+	    Context.runtime().schedule(new _ScheduleTask(self, origin, sched.event), sched.seconds);
+	}
+
+	float time() {
+	    float milliseconds = Context.runtime().now().toFloat();
+	    return milliseconds / 1000.0;
+	}
+    }
+
+    class _FakeTimeRequest {
+	Actor requester;
+	String event;
+	float happensAt;
+
+	_FakeTimeRequest(Actor requester, String event, float happensAt) {
+	    self.requester = requester;
+	    self.event = event;
+	    self.happensAt = happensAt;
+	}
+    }
+
+    @doc("Testing fake.")
+    class FakeTime extends Time, SchedulingActor {
+	float _now = 1000.0;
+	Map<long,_FakeTimeRequest> _scheduled = {};
+	MessageDispatcher dispatcher;
+
+	void onStart(MessageDispatcher dispatcher) {
+	    self.dispatcher = dispatcher;
+	}
+
+	void onMessage(Actor origin, Object msg) {
+	    Schedule sched = ?msg;
+	    _scheduled[_scheduled.keys().size()] = new _FakeTimeRequest(origin, sched.event, self._now + sched.seconds);
+	}
+
+	float time() {
+	    return self._now;
+	}
+
+	@doc("Run scheduled events whose time has come.")
+	void pump() {
+	    long idx = 0;
```

Comment:
I think it's kind-of nice if MDK compiles without quark warnings ;) (until we kill the warnings)